src/client/vegaRenderer.ts

```diff
@@ -0,0 +1,88 @@
+            const metadata: Record<typeof outputItem.mime, { embed_options?: vegaEmbed.EmbedOptions }> =
+                outputItem.metadata && typeof outputItem.metadata === 'object' && 'metadata' in outputItem.metadata
+                    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                      (outputItem.metadata as any)['metadata']
+                    : {};
+            const mode: vegaEmbed.Mode = outputItem.mime === VEGA_MIME_TYPE ? 'vega' : 'vega-lite';
+            const spec: vegaEmbed.VisualizationSpec = outputItem.json();
+            if (spec === undefined) {
+                return;
+            }
+            const mimeMetadata = metadata && outputItem.mime in metadata ? metadata[outputItem.mime] || {} : {};
+            const embedOptions: vegaEmbed.EmbedOptions =
+                typeof mimeMetadata === 'object' && mimeMetadata ? mimeMetadata.embed_options || {} : {};
+
+            if (isDarkTheme() && !embedOptions.theme && !spec.background) {
+                // Set the theme as dark only if user has not already defined a them or background color for the chart.
+                // I.e. we don't want to override user settings.
+                embedOptions.theme = 'dark';
+            }
+            // Dispose of any existing view.
+            vegaViews.get(outputItem.id)?.finalize();
+            const loader = vegaEmbed.vega.loader({
+                http: { credentials: 'same-origin' }
+            });
+            const sanitize = async (uri: string, options: unknown) => {
+                // Use the resolver for any URIs it wants to handle
+                const resolver = this._resolver;
+                if (resolver?.isLocal && resolver.isLocal(uri)) {
+                    const absPath = await resolver.resolveUrl(uri);
+                    uri = await resolver.getDownloadUrl(absPath);
+                }
+                return loader.sanitize(uri, options);
+            };
+            // Don't use `element`, else the menu options appear on the far right.
+            // Because `element` has a fixed width, and `ele` (newly created) does not.
+            const ele = document.createElement('div');
+            element.appendChild(ele);
+            const result = await vegaEmbed.default(ele, spec, {
+                actions: {
+                    export: true,
+                    compiled: false, // No point displaying these menu options if they do not work.
+                    editor: false, // No point displaying these menu options if they do not work.
+                    source: false // No point displaying these menu options if they do not work.
+                },
+                defaultStyle: true,
+                ...embedOptions,
+                mode,
+                loader: { ...loader, sanitize }
+            });
+            vegaViews.set(outputItem.id, result);
+            // Store generated image in output, send this to the extension.
```

Comment:
Left this as a reminder, will be discussing what we need to do here. Have an issue to track this work/code.